File: 2587-design-memory-allocator/2587-design-memory-allocator.py
class Allocator(object):

    def __init__(self, n):
        """
        :type n: int
        """
        self.memory = [0]*n
        

    def allocate(self, size, mID):
        """
        :type size: int
        :type mID: int
        :rtype: int
        """
        count, free = 0, 0
        
        for i in range(len(self.memory)):
            if self.memory[i] == 0:
                count += 1
            else:
                count = 0

            if count == size:
                while count:
                    self.memory[i-count+1] = mID
                    count -= 1
                return i-size+1
        return -1
        

    def freeMemory(self, mID):
        """
        :type mID: int
        :rtype: int
        """
        count = 0
        for i in range(len(self.memory)):
            if self.memory[i] == mID:
                self.memory[i] = 0
                count += 1
        return count
        


# Your Allocator object will be instantiated and called as such:
# obj = Allocator(n)
# param_1 = obj.allocate(size,mID)
# param_2 = obj.freeMemory(mID)

File: 2587-design-memory-allocator/2587-design-memory-allocator.cpp
class Allocator {
public:
    vector<int> memory;
    Allocator(int n) {
        memory.resize(n);
    }
    
    int allocate(int size, int mID) {
        int count = 0;
        int free = 0;
        
        for(int i = 0; i < memory.size(); i++){
            if(memory[i] == free) count++;
            else count = 0;

            if(count == size){
                while(count){
                    memory[i-count+1] = mID;
                    count--;
                }
                return i-size+1;
            }
        }

        return -1;
    }
    
    int freeMemory(int mID) {
        int count = 0;
        for(int i = 0; i < memory.size(); i++){
            if(memory[i] == mID) {
                memory[i] = 0;
                count++;
            }
        }
        return count;
    }
};

/**
 * Your Allocator object will be instantiated and called as such:
 * Allocator* obj = new Allocator(n);
 * int param_1 = obj->allocate(size,mID);
 * int param_2 = obj->freeMemory(mID);
 */